File: 0078_subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        output = []
        path = []
        self.dfs(nums, output, path)
        return output
        
    def dfs(self, nums, output, path):
        output.append(path)
        for i in range(len(nums)):
            self.dfs(nums[i+1:], output, path + [nums[i]])
        return

# Alternative solution

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []
        self.backtrack(result, nums, [], 0)
        return result

    def backtrack(self, result, nums, currState, start):
        result.append(currState[:])
        for i in range(start, len(nums)):
            currState.append(nums[i])
            self.backtrack(result, nums, currState, i + 1)
            currState.pop()

# Alternative solution

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = [[]]
        for num in nums:
            for i in range(len(result)):
                subset = list(result[i])
                subset.append(num)
                result.append(subset)
        return result

File: 0001-0500/0078_subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> output = new ArrayList<>();
        backtrack(nums, output, new ArrayList<>(), 0);
        return output;
    }

    public void backtrack(int[] nums, List<List<Integer>> output, List<Integer> path, int index) {
        output.add(new ArrayList<>(path));

        if (index == nums.length) {
            return;
        }

        for (int i = index; i < nums.length; i++) {
            path.add(nums[i]);
            backtrack(nums, output, path, i + 1);
            path.remove(path.size() - 1); 
        }
    }
}

// Alternative solution

class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> output = new ArrayList<>();
        output.add(new ArrayList<>());

        for (int num: nums) {
            int n = output.size();
            for (int i = 0; i < n; i++) {
                List<Integer> newList = new ArrayList<>(output.get(i));
                newList.add(num);
                output.add(newList);
            }
        }

        return output;
    }
}
